cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeJobCleaner.java

```diff
@@ -0,0 +1,95 @@
+  private static final Logger LOG = LoggerFactory.getLogger(KubeJobCleaner.class);
+  private final BatchV1Api batchV1Api;
+  private final String kubeNamespace;
+  private final String selector;
+  private final int batchSize;
+
+  KubeJobCleaner(BatchV1Api batchV1Api, String kubeNamespace, String selector, int batchSize) {
+    this.batchV1Api = batchV1Api;
+    this.kubeNamespace = kubeNamespace;
+    this.selector = selector;
+    this.batchSize = batchSize;
+  }
+
+  @Override
+  public void run() {
+    String continuationToken = null;
+    int retryCount = 10;
+    int jobDeletionCount = 0;
+    do {
+      try {
+        // Attempt to delete completed jobs. K8s current implementation only supports status.successful field selector
+        // for jobs. https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/v1/conversion.go
+        // so instead, list all the jobs in a given k8s namespace and delete completed (successful + failed) jobs one
+        // by one.
+        V1JobList jobs = batchV1Api.listNamespacedJob(kubeNamespace, null, null, continuationToken, null, selector,
+                                                      batchSize, null, null,
+                                                      (int) TimeUnit.MINUTES.toSeconds(10), null);
+        for (V1Job job : jobs.getItems()) {
+          V1JobStatus jobStatus = job.getStatus();
+          // Only attempt to delete completed jobs.
+          if (jobStatus != null && (jobStatus.getSucceeded() != null || jobStatus.getFailed() != null)) {
+            String jobName = job.getMetadata().getName();
+            V1DeleteOptions v1DeleteOptions = new V1DeleteOptions();
+            v1DeleteOptions.setPropagationPolicy("Background");
+            try {
+              // Rely on k8s garbage collector to delete dependent pods in background while job resource is deleted
+              // immediately - https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection
+              batchV1Api.deleteNamespacedJob(jobName, kubeNamespace, null, null, null, null, null, v1DeleteOptions);
+              jobDeletionCount++;
+            } catch (ApiException e) {
+              if (e.getCode() == 404) {
+                // Ignore if status code is 404, this could happen in case there is some race condition while issuing
+                // delete for the same job.
+                LOG.trace("Ignoring job deletion for job {} because job was not found.", jobName, e);
+              } else {
+                // catch the exception so that we can proceed with other job deletions.
+                LOG.warn("Failed to cleanup job resources for job {}. This attempt will be retried later.", jobName, e);
+              }
+            }
+          }
+        }
+        continuationToken = jobs.getMetadata().getContinue();
+      } catch (ApiException e) {
+        // This could happen if there was error while listing the jobs.
+        retryCount--;
```

Comment:
should we sleep for some time before retrying?

Reply:
yea thats a good idea, may be for small amount of time like 200 millis or something like that.